### PR TITLE
BUG: na_rep given precedence in to_html

### LINKED ISSUE #13828: to_html ignores na_rep when float_format set
#### Code Sample, a copy-pastable example if possible

http://stackoverflow.com/questions/23700835/pandas-to-html-no-value-representation

df = pd.DataFrame([['A',1.2225],['A',]], columns=['Group','Data'])
df.to_html(na_rep="Ted",float_format='{0:.2f}'.format)
#### Expected Output

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Group</th>
      <th>Data</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td> A</td>
      <td>1.22</td>
    </tr>
    <tr>
      <th>1</th>
      <td> A</td>
      <td> Ted</td>
    </tr>
  </tbody>
</table>

## INSTALLED VERSIONS

commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.2
setuptools: 20.3
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.2
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.5.1
pytz: 2016.2
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.6.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0

### PR DESCRIPTION
**Pull request by agraboso**
- [x] closes #13828
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

~~I know new tests are appreciated from the get-go, but it's the end of the day for me. I'll add new tests and the appropriate whatsnew entry tomorrow.~~

### CODE REVIEW

File: pandas/formats/format.py [hunk 1]
```diff
@@ -2042,7 +2042,12 @@ def get_result_as_array(self):
         """
 
         if self.formatter is not None:
-            return np.array([self.formatter(x) for x in self.values])
+            # na_rep takes precedence over the specified formatter
+            if self.na_rep is not None:
+                return np.array([self.formatter(x) if notnull(x)
+                                 else self.na_rep for x in self.values])
```
Comment (jreback) [inline pandas/formats/format.py]: use notnull on the entire array first to create a mask
format the non null values then fill the nans

see the datetime formatters for how to do this

Comment (agraboso) [inline pandas/formats/format.py]: @jreback I just looked at the datetime formatters but I don't know what you mean. Could you be more explicit? Thanks.

Comment (jreback) [inline pandas/formats/format.py]: like this (from internals.py)


    def to_native_types(self, slicer=None, na_rep=None, quoting=None,
                        **kwargs):
        """ convert to our native types format, slicing if desired """

        values = self.values
        if slicer is not None:
            values = values[:, slicer]
        mask = isnull(values)

        rvalues = np.empty(values.shape, dtype=object)
        if na_rep is None:
            na_rep = 'NaT'
        rvalues[mask] = na_rep
        imask = (~mask).ravel()

Comment (agraboso) [inline pandas/formats/format.py]: See update. Is that what you mean?

File: pandas/formats/format.py [hunk 2]
```diff
@@ -2042,7 +2042,13 @@ def get_result_as_array(self):
         """
 
         if self.formatter is not None:
-            return np.array([self.formatter(x) for x in self.values])
+            # na_rep takes precedence over the specified formatter
+            mask = isnull(self.values)
+            formatted_values = np.array([self.formatter(x)
+                                         for x in self.values])
+            if self.na_rep is not None:
```
Comment (jreback) [inline pandas/formats/format.py]: close but u now need to only iterate over the non null values

Comment (agraboso) [inline pandas/formats/format.py]: Why? IIUC and we're trying to get `na_rep` to have precedence over `formatter` (and `float_format`) iterating only over non-null values would be the wrong thing to do.

Comment (jreback) [inline pandas/formats/format.py]: tests will solves all of these questions

**Comment by agraboso** [comment]
@jreback I think I understood what you meant by iterating over non-null values. Please have a look.

**Comment by agraboso** [comment]
@jreback

**Comment by jreback** [comment]
can you rebase / update?

**Comment by agraboso** [comment]
Rebased and all green.

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche

File: pandas/formats/format.py [hunk 3]
```diff
@@ -2060,7 +2060,19 @@ def get_result_as_array(self):
         """
 
         if self.formatter is not None:
-            return np.array([self.formatter(x) for x in self.values])
+            if self.na_rep is None:
```
Comment (jreback) [inline pandas/formats/format.py]: you are repeating lots of code from right below ``format_values_with``. see if you can simplify

Comment (jreback) [inline pandas/formats/format.py]: can you update here

File: doc/source/whatsnew/v0.19.0.txt [hunk 4]
```diff
@@ -1524,6 +1524,7 @@ Bug Fixes
 - Bug in ``groupby(..).nth()`` where the group key is included inconsistently if called after ``.head()/.tail()`` (:issue:`12839`)
 - Bug in ``.to_html``, ``.to_latex`` and ``.to_string`` silently ignore custom datetime formatter passed through the ``formatters`` key word (:issue:`10690`)
 - Bug in ``DataFrame.iterrows()``, not yielding a ``Series`` subclasse if defined (:issue:`13977`)
+- Bug in ``.to_html``, ``.to_latex`` and ``.to_string`` ignoring ``na_rep`` in the presence of a ``float_format`` function. (:issue:`13911`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: can you move to 0.20.0

**Comment by jreback** [comment]
i think this got lost in the shuffle, can you update.

**Comment by agraboso** [comment]
Rebased, moved whatsnew to 0.20.0, need to look into the code duplication @jreback mentioned.

**Comment by jreback** [comment]
can you rebase / update

**Comment by jreback** [comment]
closing as stale. please ping if you'd like to continue.

**Pull request closed without merging**